src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -70,11 +71,18 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, ReportDiag
                     d.Category,
                     warningLevelOption,
                     generalDiagnosticOption,
-                    specificDiagnosticOptions);
+                    specificDiagnosticOptions,
+                    out hasPragmaSuppression);
             }
             else
             {
-                reportAction = GetDiagnosticReport(d.Severity, d.IsEnabledByDefault, d.Id, d.WarningLevel, d.Location as Location, d.Category, warningLevelOption, generalDiagnosticOption, specificDiagnosticOptions);
+                reportAction = GetDiagnosticReport(d.Severity, d.IsEnabledByDefault, d.Id, d.WarningLevel, d.Location as Location,
+                    d.Category, warningLevelOption, generalDiagnosticOption, specificDiagnosticOptions, out hasPragmaSuppression);
+            }
+
+            if (hasPragmaSuppression)
+            {
+                d = d.WithHasSourceSuppression(true);
```

Comment:
That is done as a post filtering step where we check SuppressMessageAttribute suppressions. You can search for other invocations of "WithHasSourceSuppression" in this PR.